Code for NMDC to NCBI export

sujaypatil96 opened the pull request:
# Description

Code that creates NCBI submission.xml using an NMDC slot-to-NCBI BioSample Attribute mapping file.

Fixes #503 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Tests are still to come.

**Configuration Details**: none

# Checklist:

- [x] My code follows the style guidelines of this project (have you run `black nmdc_runtime/`?)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (in `docs/` and in <https://github.com/microbiomedata/NMDC_documentation/>?)
- [ ] I have added tests that prove my fix is effective or that my feature works, incl. considering downstream usage (e.g. <https://github.com/microbiomedata/notebook_hackathons>) if applicable.
- [ ] New and existing unit and functional tests pass locally with my changes (`make up-test && make test-run`)

turbomam commented:
where did you get those checklists? I may want to use them in the future

`nmdc_runtime/site/export/ncbi_xml.py` looks like you're following good practices!

sujaypatil96 commented:
@turbomam it's the GitHub PR template in this repo: https://github.com/microbiomedata/nmdc-runtime/blob/main/.github/pull_request_template.md

sujaypatil96 commented:
CC: @pkalita-lbl if I could get your eyes on the code that I'm writing as part of this PR for a quick glance/review that would be super helpful! 🙏🏼

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -0,0 +1,295 @@
+            org=self.ncbi_submission_metadata.get("organization", ""),
+        )
+
+        biosamples_list = self.nmdc_api_client.get_biosamples_part_of_study(
+            self.nmdc_study_id
+        )
```
Similar to above, I would prefer to see this list obtained externally and passed to this method.

sujaypatil96 replied:
Good point! addressed this by writing logic in an `@op` called `get_biosamples_by_study_id()`.

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -0,0 +1,295 @@
+            nmdc_biosamples=biosamples_list,
+        )
+
+        rough_string = ET.tostring(self.root, "unicode")
+        reparsed = xml.dom.minidom.parseString(rough_string)
+        submission_xml = reparsed.toprettyxml(indent="    ", newl="\n")
```
Is there a reason for this serialize-deserialize step? Is it just to get nicely formatted XML? If so, would the `indent` function be better to use? https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.indent

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml_utils.py:
```diff
@@ -0,0 +1,118 @@
+    xml_schema_doc = etree.parse(BytesIO(xsd_content.encode("utf-8")))
+    xml_schema = etree.XMLSchema(xml_schema_doc)
+
+    if "<?xml" in xml:
+        xml_doc = etree.parse(BytesIO(xml.encode("utf-8")))
+    else:
+        xml_doc = etree.parse(StringIO(xml))
+
+    xml_doc = etree.parse(StringIO(xml))
```
What's the purpose of the if/else block for setting `xml_doc` when `xml_doc` is unconditionally set after it?

pkalita-lbl commented on nmdc_runtime/site/ops.py:
```diff
@@ -9,6 +9,8 @@
 from typing import Tuple
 from zipfile import ZipFile
+
+# import xml.etree.ElementTree as ET
```
Remove

pkalita-lbl commented on requirements/main.in:
```diff
@@ -30,6 +31,7 @@ pandas
 passlib[bcrypt]
 pymongo
 pydantic[email]>=1.10.0
+pytest-mock
```
This is a test-only dependency right? It should only be in `dev.in` if that's the case.

pkalita-lbl commented on requirements/dev.in:
```diff
@@ -11,4 +11,6 @@ pytest-cov
\ No newline at end of file
+requests-cache
+pytest-mock
+lxml
```
`lxml` doesn't need to be here if it's already in `main.in`.

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -0,0 +1,295 @@
+        reparsed = xml.dom.minidom.parseString(rough_string)
+        submission_xml = reparsed.toprettyxml(indent="    ", newl="\n")
+
+        # ============= Uncomment the following code to validate the XML against NCBI XSDs ============ #
+        # submission_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/submission.xsd?view=co"
+        # submission_xsd_validation = validate_xml(submission_xml, submission_xsd_url)
+
+        # bioproject_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/bioproject.xsd?view=co"
+        # bioproject_xsd_validation = validate_xml(submission_xml, bioproject_xsd_url)
+
+        # biosample_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/biosample.xsd?view=co"
+        # biosample_xsd_validation = validate_xml(submission_xml, biosample_xsd_url)
```
I wouldn't leave commented-out code like this in. But I would turn these into unit tests.

aclum commented on nmdc_runtime/site/repository.py:
```diff
@@ -852,6 +853,69 @@ def biosample_submission_ingest():
+                                "user": "",
+                                "organization": "",
+                            },
+                            "ncbi_bioproject_metadata": {
```
Code should be able to get this from OmicsProcessing

sujaypatil96 replied:
It could also be a slot on `Study`

sujaypatil96 replied:
Addressed this.

aclum commented on nmdc_runtime/site/repository.py:
```diff
@@ -852,6 +853,69 @@ def biosample_submission_ingest():
+                                "data_type": "",
+                                "exists": False,
+                            },
+                            "ncbi_biosample_metadata": {
```
Code should fetch this from biosample record

aclum commented on nmdc_runtime/site/repository.py:
```diff
@@ -852,6 +853,69 @@ def biosample_submission_ingest():
+                "ops": {
+                    "get_ncbi_export_pipeline_study_id": {
+                        "config": {
+                            "nmdc_study_id": "",
```
Consider adding support for a list of biosamples in case some samples have already been submitted from a study (ie for neon soil 1/3 of the samples are already in NCBI)

sujaypatil96 replied:
Sure, I can address this on a new PR once this is merged.

aclum reviewed (commented):
Add some logic to skip records or error if biosample metadata exists already on biosample_set 
'insdc_biosample_identifiers':{$exists:true,$not: {$size: 0}}
Same would be true for [insdc_experiment_identifiers](https://microbiomedata.github.io/nmdc-schema/insdc_experiment_identifiers/), the can currently be on OmicsProcessing. In Berkeley they can be on DataGeneration or DataObject.

sujaypatil96 commented on nmdc_runtime/site/export/nmdc_api_client.py:
```diff
@@ -0,0 +1,41 @@
+from dotenv import load_dotenv
+
+
+class NMDCApiClient:
```
Yup, I'm using the `mongo` resorces from `nmdc_runtime.site.resources` instead of defining a new client here.

sujaypatil96 commented:
@pkalita-lbl i think this code is good to go for the most part. I've addressed most of your code review comments in this PR, but i haven't addressed all. I was hoping to get this PR merged in because I want to avoid it from growing any more and becoming increasingly more painful to review.

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -0,0 +1,295 @@
+        descriptor = self.set_descriptor(title, description)
+        project_type = self.set_element("ProjectType")
+        project_type_submission = self.set_element(
+            "ProjectTypeSubmission", attrib={"sample_scope": "eEnvironment"}
```
Could you still confirm that `eEnvironment` is not a typo?

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -0,0 +1,431 @@
+        # submission_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/submission.xsd?view=co"
+        # validate_xml(submission_xml, submission_xsd_url)
+
+        # bioproject_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/bioproject.xsd?view=co"
```
This URL returns a 404?

sujaypatil96 replied:
Great catch! I've updated it to URLs that work now, and at the moment we are not enforcing XSD validation because the we are still figuring out what the XML should look like, so we are relying on actually submitting the produced XML to a test endpoint, visualizing it through the web portal/interface and coming back to the XML with modifications.

pkalita-lbl commented on nmdc_runtime/site/export/ncbi_xml.py:
```diff
@@ -0,0 +1,431 @@
+        # bioproject_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/bioproject.xsd?view=co"
+        # validate_xml(submission_xml, bioproject_xsd_url)
+
+        # biosample_xsd_url = "https://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/submit/public-docs/common/biosample.xsd?view=co"
```
This one is a 404 as well?

sujaypatil96 replied:
Fixed. Same comment as above.

pkalita-lbl commented on nmdc_runtime/site/ops.py:
```diff
@@ -724,6 +735,33 @@ def export_json_to_drs(
+)
+def ncbi_submission_xml_asset(context: OpExecutionContext, data: str):
+    filename = "ncbi_submission.xml"
+    file_path = os.path.join(context.instance.storage_directory(), filename)
```
I haven't seen this `storage_directory` function before. Is there any kind of documentation on it? I can't seem to find any.

sujaypatil96 replied:
Good catch again!

So some context as to what I'm trying to do/experiment with here. A hard requirement that came from @aclum was that we should be able to download the content, in this case, XML content that is rendered through the Dagit interface as a file on our local system.

This `storage_directory()` is part of https://docs.dagster.io/deployment/dagster-instance#local-artifact-storage which we can use to store files. Then those files need to be allowed/able to be "served up" through Dagit somehow. I'm still trying to iron out the kinks with respect to that detail, but do let me know if you have any ideas.

pkalita-lbl reviewed (approved):
There's still some room for clean up but I think this is good as a first pass
